XIPs/xip-31-message-history.md

```diff
@@ -0,0 +1,411 @@
+
+For mobile apps already handling push notifications, `MessageHistoryBackupRequest`s would become a special case of push notification handling.
+
+#### Flow for Backup Requesters
+
+1. Get a list of all other installations associated with their blockchain account. Ignore any installations that have been revoked
+2. Generate a `MessageHistoryBackupRequest` for each installation (see below for Protobuf spec) and store in the local database
+3. Send the `MessageHistoryBackupRequest` messages to the normal inbound messaging topic for each installation
+4. Wait for a response from any of the Message Backup Providers
+5. For each `MessageHistoryBackupResponse`
+
+   5a. Ensure there is a `MessageHistoryBackupRequest` with a matching `requestId` stored in the database. If not, ignore.
+
+   5b. Download the file from the `backupUrl` and decrypt using the credentials provided in the `MessageHistoryBackupResponse`.
+
+   5c. Load each message into the local database, ignoring any duplicate messages.
+
+   5d. Delete the `MessageHistoryBackupResponse` and all associated credentials.
+
+   5e. Set the status of the `MessageHistoryBackupRequest` to `Applied`.
+
+#### Flow for Message Backup Providers
+
+1. Receive a `MessageHistoryBackupRequest` as part of the normal message receiving flow.
+1. Retrieve the contact for the installation that sent the message and ensure it has not been revoked. If revoked, ignore.
+1. Ensure the installation that sent the message has a contact signed by the same blockchain account as the current user. If not, ignore.
+1. Convert all messages in the local database into a Message Backup File
+1. Generate ephemeral encryption key, salt, and nonce. Encrypt the file using these keys.
+1. Upload the file to the `backupStorageProviderUploadUrl` from the request.
+1. Reply to the message with a `MessageHistoryBackupResponse` containing the encryption details and the `backupUrl` provided by the Backup Storage Provider.
+1. Delete the `MessageHistoryBackupRequest`, the local database dump, and all the encryption keys.
+
+#### End-to-end flow
+
+![Sequence diagram illustrating the end-to-end flow for message backup providers|690x410](./assets/xip-31/message-backup-providers-flow.png)
+
+### Backup Account Files
+
+Backup account files are designed as a solution for users who have lost access to all of their XMTP apps. To create a Backup Account File, an app will generate a new set of XMTP keys and register the installation on the XMTP network. The installation contact published to the network will be set with a type of `backup`, which instructs all well-behaved clients to disregard any messages from this installation other than `MessageHistoryBackupResponse`s. This means that a compromised backup installation cannot be used to send messages on the user's behalf, which makes these files a less valuable target for attack.
```

Comment:
i didn't see any mention of a `backup` contact in the code. is this something we want to pursue? @neekolas

Reply:
It's something we're shelving for now. Nothing is preventing us from adding the backup contact concept later. From the SDK perspective a "backup installation" is just another installation and the history request system would work the same. All the complexity lives at the app level.